filesystem/azure.go

```diff
@@ -0,0 +1,208 @@
+}
+
+//AzureClient -> OpenWriteCloser function
+// If not exist, Create corresponding Container and blob.
+// At present, AzureFile.Write has a capacity restriction(10 * 1024 * 1024 bytes). 
+// I will implent unlimited version in the future.
+func (c *AzureClient) OpenWriteCloser(name string) (io.WriteCloser, error) {
+	exist, err := c.Exists(name)
+	if err != nil {
+		return nil, err
+	}
+	containerName, blobName, err := convertToAzurePath(name)
+	if err != nil {
+		return nil, err
+	}
+	if !exist {
+		_, err := c.blobClient.CreateContainerIfNotExists(containerName, storage.ContainerAccessTypeBlob)
+		if err != nil {
+			return nil, err
+		}
+		err = c.blobClient.CreateBlockBlob(containerName, blobName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &AzureFile{
+		path  : name,
+		logger : log.New(os.Stdout, "", log.Lshortfile|log.LstdFlags),
+		client : c.blobClient,
+	}, nil
+}
+
+func (f *AzureFile) Write(b []byte) (int, error) {
+	cnt, blob, err := convertToAzurePath(f.path)
+	if err != nil {
+		return 0, nil
+	}
+	blockList, err := f.client.GetBlockList(cnt, blob, storage.BlockListTypeAll)
+	blocksLen := len(blockList.CommittedBlocks) + len(blockList.UncommittedBlocks)
+	blockId := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%011d\n", blocksLen - 1)))
+	err = f.client.PutBlock(cnt, blob, blockId, b)
+	blockList, err = f.client.GetBlockList(cnt, blob, storage.BlockListTypeAll)
+	amendList := []storage.Block{}
+	for _, v := range blockList.CommittedBlocks {
+		amendList = append(amendList, storage.Block{v.Name, storage.BlockStatusCommitted})	
+	}
+	for _, v := range blockList.UncommittedBlocks {
+		amendList = append(amendList, storage.Block{v.Name, storage.BlockStatusUncommitted})
+	}
+	err = f.client.PutBlockList(cnt, blob, amendList)
+	if (err != nil) {
+		return 0, nil
+	}
+	return 0, nil
+}
+
+func (f *AzureFile) Close() error {
+	return nil
+}
+
+// AzureClient -> Glob function
+// Syntax : Adf{1,3}?ee*/ytsd.*
+// Follow regexp syntax except "/"
+func (c *AzureClient) Glob(pattern string) (matches []string, err error) {
+	afterSplit := strings.Split(pattern, "/")
+	cntPattern, blobPattern := afterSplit[0], afterSplit[1]
+	if len(afterSplit) != 2 {	
+		return nil, fmt.Errorf("Glob pattern should follow the Syntax")
+	}
+	resp, err := c.blobClient.ListContainers(storage.ListContainersParameters{Prefix: ""})
+	if err != nil {
+		return nil, err
+	}
+	for _, cnt := range resp.Containers {
+		if match, err := regexp.MatchString(cntPattern, cnt.Name); match && err == nil {
+			resp, err := c.blobClient.ListBlobs(cnt.Name, storage.ListBlobsParameters{
+				Marker:     ""})
+			if err != nil {
+				return nil, err
+			}
+			for _, v := range resp.Blobs {
+				if match, err := regexp.MatchString(blobPattern, v.Name); match && err == nil {
+					matches = append(matches, cnt.Name + "/" + v.Name)
+				}
+			}
+		}
+	}
+	return matches, nil
+}
+
+func randString(n int) string {
```

Comment:
While we can checking, please add test, follow Hongchao's hdfs_test.go

Xiaoyun

On Sun, Mar 22, 2015 at 10:50 PM, Zifei Tong <notifications@github.com>
wrote:

> In filesystem/azure.go
> <https://github.com/taskgraph/taskgraph/pull/40#discussion_r26915283>:
>
> > +			resp, err := c.blobClient.ListBlobs(cnt.Name, storage.ListBlobsParameters{
> > +				Marker:     ""})
> > +			if err != nil {
> > +				return nil, err
> > +			}
> > +			for _, v := range resp.Blobs {
> > +				if match, err := regexp.MatchString(blobPattern, v.Name); match && err == nil {
> > +					matches = append(matches, cnt.Name + "/" + v.Name)
> > +				}
> > +			}
> > +		}
> > +	}
> > +	return matches, nil
> > +}
> > +
> > +func randString(n int) string {
>
> Is this function used anywhere?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/taskgraph/taskgraph/pull/40/files#r26915283>.
>